kythe/docs/rfc/2966.md

```diff
@@ -0,0 +1,331 @@
+    }
+    ```
+
+-   `AnalyzerRequest`
+
+    ```
+    {
+      "language": <language-label>
+    }
+    ```
+
+-   `AnalyzeReply`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "unit": <compilation>
+    }
+    ```
+
+-   `FileRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "path": <file-path>,
+      "digest": <digest>
+    }
+    ```
+
+-   `FileReply`
+
+    ```
+    {
+      "path": <file-path>,
+      "digest": <digest>,
+      "data": <bytes>
+    }
+    ```
+
+-   `OutRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "output": [<bytes>, ...],  -- may be omitted if empty
+      "entries": [<entry>, ...]  -- may be omitted if empty
+    }
+    ```
+
+-   `LogRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "message": <diagnostic>
+    }
+    ```
+
+The basic types referenced above are:
+
+-   `<analysis-id>`: an integer ID assigned by the server (JSON: number).
+-   `<bytes>`: a base64-encoded byte string (JSON: string).
+-   `<compilation>`: a `kythe.proto.CompilationUnit` message (JSON: object).
+-   `<diagnostic>`: a diagnostic message (JSON: string).
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
+-   `<entry>`: a `kythe.proto.Entry` message (JSON: object).
+-   `<file-path>`: a file path (JSON: string).
+-   `<language-label>`: a Kythe language label (JSON: string).
+-   `<protocol>`: a protocol version label (JSON: string).
+
+Where the description refers to a protocol buffer message, the message is
+encoded using the canonical
+[JSON mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
+for the named message type.
+
+### Methods
+
+A driver defines the following JSON-RPC methods:
+
+-   `init(InitRequest) → InitReply`
+
+    The `init` method is a protocol handshake, and must be the first method
+    called by the analyzer upon startup. The request specifies the protocol
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. The environment assigns a unique integer analysis ID to each
+    request, distinct from any other ID currently assigned to that analyzer.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":2,"method":"analyze","params":{"language":"go"}}
+        « {"jsonrpc":"2.0","id":2","result":{"id":101,"unit":{...}}}
+
+-   `file(FileRequest) → FileReply`
```

Comment:
I would like some more context about parallel file requests in this section.

Can files be requested in parallel?  Is that decision left to the driver implementation?  Does the client need to use a JSON-RPC batch request or can individual requests be parallel? If files can be requested in parallel, can their contents be returned out of request order?

Reply:
I added an example with batch file requests, and added an explicit requirement that the server support both batch and concurrent in-flight requests from the client.